File: web/pos/src/app/pos/lib/topping-pricing.ts
/**
 * Frontend port of `App\Services\Topping\ToppingPricingService`.
 *
 * Pure pricing math for topping selections. Used as a live preview in
 * `ProductOptionsDialog` so the running total updates without a server
 * roundtrip on every checkbox toggle.
 *
 * **Source of truth is the PHP service** — the wire payload always reaches
 * `CustomerOrderService::addItems` and the persisted `topping_subtotal`
 * comes from the PHP implementation. This file MUST mirror the PHP
 * algorithm exactly. Parity tests live next to this file (Vitest) and
 * mirror the Pest test fixtures in `tests/Unit/Topping/`.
 *
 * Plan 015 / OQ-1: free_up_to_n waives the MOST EXPENSIVE N selections
 * (Toast/Square industry default, customer-friendly framing). Reverse
 * the sort comparator to flip to cheapest-first — and update the PHP
 * service in the same change.
 */

import type {
  ShopMenuToppingGroup,
  ToppingPriceStrategy,
} from "@/app/pos/types";

/**
 * One topping selection during dialog interaction. Mirrors PHP shape.
 *
 * `unitPrice` is resolved client-side by reading
 * `ShopMenuToppingItemSku.extra_price` for the chosen `productSkuId`,
 * with NULL row fallback for simple toppings.
 */
export interface PricedSelection {
  toppingGroupItemId: string;
  productSkuId: string;
  quantity: number;
  unitPrice: number;
}

export interface PricingBreakdownEntry {
  toppingGroupItemId: string;
  productSkuId: string;
  unitPrice: number;
  charged: boolean;
}

export interface PricingResult {
  /** Per-unit topping cost — multiply by line.quantity in the caller. */
  toppingSubtotal: number;
  /** One entry per individual unit (after expanding by qty). */
  breakdown: PricingBreakdownEntry[];
}

/**
 * Compute per-unit topping subtotal for selections within ONE group.
 * Caller must filter by group before calling — strategy/free_quantity
 * is per-group, mixing groups gives wrong math.
 */
export function priceLine(
  selections: PricedSelection[],
  group: Pick<
    ShopMenuToppingGroup,
    "price_strategy" | "free_quantity"
  >,
): PricingResult {
  if (selections.length === 0) {
    return { toppingSubtotal: 0, breakdown: [] };
  }

  // Expand by quantity into individual units. free_up_to_n waives N
  // units, not N selections — "Mayo ×3" contributes 3 units to the sort.
  const units = selections.flatMap((s) =>
    Array.from({ length: Math.max(1, s.quantity) }, () => ({
      toppingGroupItemId: s.toppingGroupItemId,
      productSkuId: s.productSkuId,
      unitPrice: s.unitPrice,
    })),
  );

  const strategy: ToppingPriceStrategy = group.price_strategy;
  if (strategy === "flat") {
    return priceFlat(units);
  }
  if (strategy === "free_up_to_n") {
    return priceFreeUpToN(units, group.free_quantity ?? 0);
  }
  // Unknown strategy — defensively treat as flat. Server-side will reject.
  return priceFlat(units);
}

/**
 * Sum across multiple groups for one line. Pass `groupedSelections`
 * keyed by `topping_group_id` with each bucket's group config + the
 * subset of selections targeting that group. Mirrors PHP
 * `priceLineAcrossGroups`.
 */
export function priceLineAcrossGroups(
  groupedSelections: Record<
    string,
    {
      group: Pick<ShopMenuToppingGroup, "price_strategy" | "free_quantity">;
      selections: PricedSelection[];
    }
  >,
): PricingResult {
  let total = 0;
  const breakdown: PricingBreakdownEntry[] = [];

  for (const bucket of Object.values(groupedSelections)) {
    const result = priceLine(bucket.selections, bucket.group);
    total += result.toppingSubtotal;
    for (const entry of result.breakdown) {
      breakdown.push(entry);
    }
  }

  return { toppingSubtotal: round2(total), breakdown };
}

interface Unit {
  toppingGroupItemId: string;
  productSkuId: string;
  unitPrice: number;
}

function priceFlat(units: Unit[]): PricingResult {
  let subtotal = 0;
  const breakdown: PricingBreakdownEntry[] = [];
  for (const unit of units) {
    subtotal += unit.unitPrice;
    breakdown.push({ ...unit, charged: true });
  }
  return { toppingSubtotal: round2(subtotal), breakdown };
}

function priceFreeUpToN(units: Unit[], freeQuantity: number): PricingResult {
  if (freeQuantity <= 0) {
    return priceFlat(units);
  }

  // Stable sort by unitPrice DESC, ties resolved by original insertion
  // index. Stable so parity tests against PHP get deterministic
  // breakdowns when prices repeat.
  const indexed = units.map((u, idx) => ({ ...u, _idx: idx }));
  indexed.sort((a, b) => {
    const cmp = b.unitPrice - a.unitPrice;
    return cmp !== 0 ? cmp : a._idx - b._idx;
  });

  let subtotal = 0;
  let waived = 0;
  const chargedByIdx = new Map<number, boolean>();
  for (const u of indexed) {
    const charged = waived >= freeQuantity;
    if (charged) {
      subtotal += u.unitPrice;
    } else {
      waived++;
    }
    chargedByIdx.set(u._idx, charged);
  }

  // Map back to original order so receipts/log entries follow the
  // customer's pick order, not the sort order.
  const breakdown: PricingBreakdownEntry[] = units.map((u, idx) => ({
    toppingGroupItemId: u.toppingGroupItemId,
    productSkuId: u.productSkuId,
    unitPrice: u.unitPrice,
    charged: chargedByIdx.get(idx) ?? true,
  }));

  return { toppingSubtotal: round2(subtotal), breakdown };
}

/**
 * Currency rounding to 2 decimals. Decimal(15,2) on backend; PHP uses
 * `round(..., 2)` (half-away-from-zero). JS `Math.round` differs for
 * 0.5 ties — to keep parity with PHP we apply the standard
 * (Math.round((x + Number.EPSILON) * 100) / 100) idiom.
 */
function round2(value: number): number {
  return Math.round((value + Number.EPSILON) * 100) / 100;
}

File: workstation/internal/service/topping_pricing.go
package service

import (
	"database/sql"
	"sort"
)

// Port of pos-web/src/app/pos/lib/topping-pricing.ts (priceLineAcrossGroups →
// priceLine → priceFreeUpToN → priceFlat), itself a port of Cloud's
// App\Services\Topping\ToppingPricingService. This MUST produce numbers
// identical to the TS/PHP implementation so the LAN cart line matches the
// pos-web running total and Cloud's persisted topping_subtotal (BR-OI06).
//
// All prices here are integer minor units (extra_price / override_price /
// unit_price are INTEGER columns and int fields). The TS round2() rounds to 2
// decimals half-away-from-zero; with integer inputs it is the identity, so the
// integer sum below is exact parity — no float step to diverge.
//
// Plan 015 / OQ-1: free_up_to_n waives the MOST EXPENSIVE N units (Toast/Square
// industry default). Reverse the DESC comparator to flip to cheapest-first —
// and update topping-pricing.ts + the PHP service in the same change.

// pricedTopping is the minimal per-selection shape the algorithm needs.
type pricedTopping struct {
	ToppingGroupID string
	UnitPrice      int
	Quantity       int
}

// toppingGroupPricing carries the per-group strategy config from
// pos_topping_groups. PriceStrategy: "flat" | "free_up_to_n"; anything else
// (including "" for an unsynced group) is treated as flat.
type toppingGroupPricing struct {
	PriceStrategy string
	FreeQuantity  int // pos_topping_groups.free_quantity; NULL → 0
}

// priceLineAcrossGroups sums the per-unit topping subtotal for one order line,
// bucketing selections by topping_group_id and applying each group's strategy.
// Mirrors the TS function of the same name. Free quotas never leak across
// groups. First-seen group order is preserved for deterministic iteration
// (parity with JS Object.values insertion order); it does not affect the sum.
func priceLineAcrossGroups(toppings []pricedTopping, groups map[string]toppingGroupPricing) int {
	order := make([]string, 0)
	buckets := make(map[string][]pricedTopping)
	for _, t := range toppings {
		if _, ok := buckets[t.ToppingGroupID]; !ok {
			order = append(order, t.ToppingGroupID)
		}
		buckets[t.ToppingGroupID] = append(buckets[t.ToppingGroupID], t)
	}
	total := 0
	for _, gid := range order {
		total += priceLine(buckets[gid], groups[gid])
	}
	return total
}

// priceLine prices selections within ONE group. Caller must filter by group.
func priceLine(sel []pricedTopping, group toppingGroupPricing) int {
	if len(sel) == 0 {
		return 0
	}
	// Expand by quantity into individual units — free_up_to_n waives units, not
	// selections ("Mayo ×3" contributes 3 units to the sort).
	units := make([]int, 0, len(sel))
	for _, s := range sel {
		q := max(1, s.Quantity) // TS Math.max(1, s.quantity)
		for range q {
			units = append(units, s.UnitPrice)
		}
	}
	switch group.PriceStrategy {
	case "free_up_to_n":
		return priceFreeUpToN(units, group.FreeQuantity)
	default:
		// "flat" and any unknown/unsynced strategy price every unit.
		return priceFlat(units)
	}
}

func priceFlat(units []int) int {
	sum := 0
	for _, u := range units {
		sum += u
	}
	return sum
}

// priceFreeUpToN waives the freeQuantity MOST EXPENSIVE units, charging the
// rest. Mirrors the TS priceFreeUpToN (stable DESC sort, waive first N).
func priceFreeUpToN(units []int, freeQuantity int) int {
	if freeQuantity <= 0 {
		return priceFlat(units)
	}
	sorted := make([]int, len(units))
	copy(sorted, units)
	// Stable DESC sort — sort.SliceStable preserves input order for equal keys,
	// replicating the TS "_idx" tiebreak without an explicit index field.
	sort.SliceStable(sorted, func(i, j int) bool { return sorted[i] > sorted[j] })

	sum := 0
	waived := 0
	for _, u := range sorted {
		if waived >= freeQuantity {
			sum += u
		} else {
			waived++
		}
	}
	return sum
}

// loadToppingGroupPricing fetches price_strategy + free_quantity for the
// distinct topping_group_ids on a line, from the synced pos_topping_groups
// replica. NULL free_quantity → 0. A group missing from the replica yields the
// zero value, which priceLine treats as flat (safe default). Must run AFTER
// resolveToppingSnapshot so each ToppingInput.ToppingGroupID is hydrated.
func (e *OrderEngine) loadToppingGroupPricing(toppings []ToppingInput) map[string]toppingGroupPricing {
	out := map[string]toppingGroupPricing{}
	for _, t := range toppings {
		gid := t.ToppingGroupID
		if gid == "" {
			continue
		}
		if _, seen := out[gid]; seen {
			continue
		}
		var strategy string
		var freeQty sql.NullInt64
		if err := e.db.QueryRow(
			`SELECT price_strategy, free_quantity FROM pos_topping_groups WHERE id = ?`, gid,
		).Scan(&strategy, &freeQty); err == nil {
			g := toppingGroupPricing{PriceStrategy: strategy}
			if freeQty.Valid {
				g.FreeQuantity = int(freeQty.Int64)
			}
			out[gid] = g
		}
	}
	return out
}

// menuProductIDForSku resolves the pos_menu_products.id (the tier-1 topping
// override key) for a parent line's ProductSku. A product can be attached to
// several menus, so we mirror the menu-list ordering pos-web renders from
// (published/active menus, sort_order then name) and take the first — the same
// tile the operator tapped. Returns "" when the SKU is on no published menu, in
// which case the caller skips tier-1 and behaves exactly as before this fix.
func (e *OrderEngine) menuProductIDForSku(parentSkuID string) string {
	if parentSkuID == "" {
		return ""
	}
	// plan-056 — `COALESCE(ov.is_active, mp.is_active)`, not `mp.is_active`.
	//
	// The gate has to mean "is this dish on the menu the operator is looking
	// at", and after plan-056 a dish can be off because of a LOCAL toggle that
	// has not reached Cloud yet. Reading the replica column alone would keep
	// choosing a just-turned-off menu_product as the tier-1 source of topping
	// prices — a wrong add-on price on a real bill, from a row the screen no
	// longer shows.
	var menuProductID string
	_ = e.db.QueryRow(`
		SELECT mp.id
		FROM pos_product_skus ps
		JOIN pos_menu_products mp ON mp.product_id = ps.product_id
		JOIN pos_menus m          ON m.id = mp.menu_id
		LEFT JOIN pos_menu_availability_overrides ov
		       ON ov.entity_type = 'menu_product' AND ov.entity_id = mp.id
		WHERE ps.id = ?
		  AND COALESCE(ov.is_active, mp.is_active) = 1
		  AND m.status IN ('published','Published','active','Active')
		ORDER BY m.sort_order, m.name
		LIMIT 1`, parentSkuID).Scan(&menuProductID)
	return menuProductID
}
